// fnts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.identity = exports.ternary = exports.rethrow = exports.compose = exports.inject = exports.match = exports.curry = exports.pipe = exports.tap = void 0;
var tap_1 = require("./tap");
Object.defineProperty(exports, "tap", { enumerable: true, get: function () { return tap_1.default; } });
var pipe_1 = require("./pipe");
Object.defineProperty(exports, "pipe", { enumerable: true, get: function () { return pipe_1.default; } });
var curry_1 = require("./curry");
Object.defineProperty(exports, "curry", { enumerable: true, get: function () { return curry_1.default; } });
var guard_1 = require("./guard");
Object.defineProperty(exports, "match", { enumerable: true, get: function () { return guard_1.default; } });
var inject_1 = require("./inject");
Object.defineProperty(exports, "inject", { enumerable: true, get: function () { return inject_1.default; } });
var compose_1 = require("./compose");
Object.defineProperty(exports, "compose", { enumerable: true, get: function () { return compose_1.default; } });
var rethrow_1 = require("./rethrow");
Object.defineProperty(exports, "rethrow", { enumerable: true, get: function () { return rethrow_1.default; } });
var ternary_1 = require("./ternary");
Object.defineProperty(exports, "ternary", { enumerable: true, get: function () { return ternary_1.default; } });
var identity_1 = require("./identity");
Object.defineProperty(exports, "identity", { enumerable: true, get: function () { return identity_1.default; } });
