// fnts/either/operators/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.second = exports.first = exports.bimap = exports.bifoldMap = exports.bindr = exports.bifold = exports.bifoldr = exports.bifoldl = void 0;
var bifold_1 = require("./bifold");
Object.defineProperty(exports, "bifoldl", { enumerable: true, get: function () { return bifold_1.bifoldl; } });
Object.defineProperty(exports, "bifoldr", { enumerable: true, get: function () { return bifold_1.bifoldr; } });
Object.defineProperty(exports, "bifold", { enumerable: true, get: function () { return bifold_1.default; } });
__exportStar(require("./guards"), exports);
var bindr_1 = require("./bindr");
Object.defineProperty(exports, "bindr", { enumerable: true, get: function () { return bindr_1.default; } });
var bifold_map_1 = require("./bifold-map");
Object.defineProperty(exports, "bifoldMap", { enumerable: true, get: function () { return bifold_map_1.default; } });
var bimap_1 = require("./bimap");
Object.defineProperty(exports, "bimap", { enumerable: true, get: function () { return bimap_1.default; } });
Object.defineProperty(exports, "first", { enumerable: true, get: function () { return bimap_1.first; } });
Object.defineProperty(exports, "second", { enumerable: true, get: function () { return bimap_1.second; } });
